crmsh/ui_corosync.py

```diff
@@ -47,6 +49,190 @@ def _diff_nodes(args):
+            print('\n  Options:')
+            for name, value in dataclasses.asdict(link).items():
+                if name == 'linknumber' or name == 'nodes':
+                    continue
+                if value is None:
+                    continue
+                print(f'    {name}:\t{value}')
+            print('')
+        # TODO: show link status
+
+    def do_update(self, context, *argv):
+        lm = corosync.LinkManager.load_config_file()
+        if lm.totem_transport() != 'knet':
+            logger.error('Corosync is not using knet transport')
+            return False
+        try:
+            args = LinkArgumentParser().parse(True, True, argv)
+        except LinkArgumentParser.SyntaxException as e:
+            logger.error('%s', str(e))
+            print('Usage: link update <linknumber> [<node>=<addr> ...] [options <option>=[<value>] ...] ', file=sys.stderr)
+            return False
+        self._validate_node_addresses(dict(args.nodes))
+        lm.update_link(args.linknumber, args.options)   # this also verifies if args.linknumber is valid
+        nodes = lm.links()[args.linknumber].nodes
+        node_addresses: dict[int, str] = dict()
+        for name, addr in args.nodes:
+            nodeid = next((x.nodeid for x in nodes if x.name == name), -1)
+            if nodeid == -1:
+                logger.error(f'Unknown node {name}.')
+            node_addresses[nodeid] = addr
+        if not args.nodes and not args.options:
+            logger.info("Nothing is updated.")
+        else:
+            lm.write_config_file(
+                lm.update_node_addr(args.linknumber, node_addresses)
+            )
+            logger.info("Use \"crm corosync diff\" to show the difference")
+            logger.info("Use \"crm corosync push\" to sync")
+
+    def do_add(self, context, *argv):
+        lm = corosync.LinkManager.load_config_file()
+        if lm.totem_transport() != 'knet':
+            logger.error('Corosync is not using knet transport')
+            return False
+        try:
+            args = LinkArgumentParser().parse(False, False, argv)
+        except LinkArgumentParser.SyntaxException as e:
+            logger.error('%s', str(e))
+            print('Usage: link add <node>=<addr> ... [options <option>=<value> ...] ', file=sys.stderr)
+            return False
+        self._validate_node_addresses(dict(args.nodes))
+        nodes = lm.links()[0].nodes
+        node_addresses: dict[int, str] = dict()
+        for name, addr in args.nodes:
+            nodeid = next((x.nodeid for x in nodes if x.name == name), -1)
+            if nodeid == -1:
+                logger.error(f'Unknown node {name}.')
+            node_addresses[nodeid] = addr
+        lm.write_config_file(
+            lm.add_link(node_addresses, args.options)
+        )
+        logger.info("Use \"crm corosync diff\" to show the difference")
+        logger.info("Use \"crm corosync push\" to sync")
```

Comment:
`do_update` and `do_add` method have some repetition codes, it'd better to extract common part